### ISSUE TITLE
Update text in Sort an Array Alphabetically using the sort Method Challenge

### ISSUE THREAD

**Issue by ezioda004** [freeCodeCamp/freeCodeCamp#25829]
**Describe your problem and how to reproduce it:**
This challenge has an example of sorting like this:
```js
function reverseAlpha(arr) {
  return arr.sort(function(a, b) {
    return a < b;
  });
}
reverseAlpha(['l', 'h', 'z', 'b', 's']);
````
This example uses a bad compare function. In recent changes to Chrome the following example no longer properly sorts the array:
```js
reverseAlpha(['l', 'h', 'z', 'b', 's']);
/* output
["l", "h", "z", "b", "s"] - in latest chrome 70
and not 
['z', 's', 'l', 'h', 'b'] - this is shown in the example
*/
```
The compare function needs to return a number `1, 0 or -1` to be able to sort an array properly.

### Resolve:

This can be fixed by changing the compare function to return a number:
```js
function reverseAlpha(arr) {
  return arr.sort(function(a, b) {
    return a < b ? 1 : -1;
  });
}
reverseAlpha(['l', 'h', 'z', 'b', 's']); //  ["z", "s", "l", "h", "b"]
```
**Add a Link to the page with the problem:**
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/functional-programming/sort-an-array-alphabetically-using-the-sort-method/

**Tell us about your browser and operating system:**
* Browser Name: Chrome
* Browser Version: 70
* Operating System: Windows 10

**Comment by RandellDawson** [comment]
> The compare function needs to return a number 1, 0 or -1 to be able to sort an array properly.
```
return a < b ? 1 : -1;
```
The above return statement only returns 1 or -1.  To preserve the original order for elements which have the same value, you need to test for the condition where they are equal.  For example, to preserve an elements original order when equal, you would write:
```
return a === b ? 0 : a < b ? 1 : -1;
```
While in this particular example, you may not be able to see the difference, I will show a different example below, where the difference can be seen.

Let's say you have the following array of objects:
```
const letters = [
  { letter: 'l', pos: 0 },
  { letter: 'b', pos: 1 },
  { letter: 'h', pos: 2 },
  { letter: 'z', pos: 3 },
  { letter: 'b', pos: 4 },
  { letter: 's', pos: 5 }
]
```
The 2nd and fifth elements have the same letter, but they are in different positions.  Let's say you want to sort the objects in alphabetical order **and** preserve the original sort order of the elements with the same letter (in this case those with letter 'b').  This means, after the sort, letters should look like:
```
[ { letter: 'b', pos: 1 },
  { letter: 'b', pos: 4 },
  { letter: 'h', pos: 2 },
  { letter: 'l', pos: 0 },
  { letter: 's', pos: 5 },
  { letter: 'z', pos: 3 } ]
```  
If you just create the following sort function without specifying a return value of 0 when the letters are equal, it would not sort correctly in the new version of Chrome:
```
const letters = [
  { letter: 'l', pos: 0 },
  { letter: 'b', pos: 1 },
  { letter: 'h', pos: 2 },
  { letter: 'z', pos: 3 },
  { letter: 'b', pos: 4 },
  { letter: 's', pos: 5 }
]
letters.sort(function(a, b) {
  return a.letter > b.letter ? 1 : -1;
});
```
The above produces:
```
[ { letter: 'b', pos: 4 },
  { letter: 'b', pos: 1 },
  { letter: 'h', pos: 2 },
  { letter: 'l', pos: 0 },
  { letter: 's', pos: 5 },
  { letter: 'z', pos: 3 } ]
```
instead of the correct version:
```
[ { letter: 'b', pos: 1 },
  { letter: 'b', pos: 4 },
  { letter: 'h', pos: 2 },
  { letter: 'l', pos: 0 },
  { letter: 's', pos: 5 },
  { letter: 'z', pos: 3 } ]
```
To preserve the original sort order for objects with the same letter, you must use the following sort function:
```
letters.sort(function (a, b) {
    return a.letter === b.letter ? 0 : a.letter > b.letter ? 1 : -1;
  });
```
I think the example should reflect the better practice of specifying a return value of 0 for elements which are equal in value.  This matches the actual requirements of the compare function of returning a value of 0, -1, or 1.

**Comment by rpjayasekara** [comment]
I am interested working on this issue

**Comment by ezioda004** [comment]
@RandellDawson Although the example used in the original text doesn't require ordering I agree adding `0` would be fine. 
As much as I like ternaries, I'm not in favor of nested ternaries. As a general rule they [shouldnt be nested](https://github.com/airbnb/javascript#comparison--nested-ternaries). If possible I'd keep the text challenge clean and readable. Maybe the good ol' if statements are fine here:
```js
function reverseAlpha(arr) {
  return arr.sort(function(a, b) {
    if (a === b) {
      return 0;
   }
     else if (a < b) {
      return 1;
    }
    else {
      return -1;
    }
  });
}
```

**Comment by RandellDawson** [comment]
Let's see what @ezioda004 says after my last reply, before we assign it to you.

**Comment by Nirajn2311** [comment]
Yeah, even I think it is better to make it in the form of `if else statements` as many people still get confused with nested ternaries.

**Comment by utkarsh-raj** [comment]
I would take a different perspective here. For the purposes of the tutorial, I suggest that nested ternaries are the way to go. If people do get confused, this will encourage them to look up their doubt and they will be better versed with ternaries for later.
I have opened a PR with the edit implementing nested ternaries.

**Comment by ezioda004** [comment]
@RandellDawson On second thought, and the fact we already have a nested ternary lesson in basic JS section. I think it's fine to use a nested ternary.

@utkarsh-raj Good point as well.

**Comment by RandellDawson** [comment]
> I have opened a PR with the edit implementing nested ternaries.

Make sure to reference that it closes this issue # if you have not already done so.

**Comment by Nirajn2311** [comment]
@utkarsh-raj just remember to ask first before you create a PR as there might be other people waiting to do one, like @rpjayasekara had asked here before if he could work on it?

@rpjayasekara sorry about not getting to do a PR for this issue, there are other issues you might take a look at.

**Comment by utkarsh-raj** [comment]
@Nirajn2311 I will take care of this in future PRs. Apologies for my mistake @rpjayasekara .

**Comment by Amjadamjad** [comment]
#25978

**Comment by Amjadamjad** [comment]
#25829

**Comment by Amjadamjad** [comment]
#25829
#25978

**Comment by RandellDawson** [comment]
@Amjadamjad What are these PRs you have referenced?

**Comment by Nirajn2311** [comment]
@RandellDawson could be spam.

**Comment by drecali** [comment]
The sample code on the FCC website is still the old version. I noticed on the challenge's GitHub page the sample code is different than what I saw on the live FCC website. 

I did this challenge on December 19, 2018. The sample code on the FCC website website is the old version (different from the GitHub version). I thought that the GitHub version of the challenge was reflected on the FCC site. Does this happen at regular intervals, or do updated challenges need to be pushed live to the site?

As a side note, the sample code on the GitHub page and the live FCC site doesn't work in Chrome 70+ for reasons addressed above. May I update the sample code to a version that works? In this case, the description would also need updating to explain the code. Should I make the changes as a pull request? Sorry, I'm really new to FCC and Open Source contributing in general. I just want to make sure I do things in the most efficient way.

I propose the following (the opposite of the reverse alphabetical sort suggested above):
```js
function alphabeticalOrder(arr) {
  return arr.sort(function(a, b) {
    return a === b ? 0 : a > b ? 1 : -1;
  });
}
alphabeticalOrder(["a", "d", "c", "a", "z", "g"]);
```

The old version that doesn't work in Chrome 70+:
```js
function alphabeticalOrder(arr) {
  return arr.sort(function(a, b) {
    return a - b;
  });
}
alphabeticalOrder(["a", "d", "c", "a", "z", "g"]);
```

Here's the screenshot: ![screenshot](http://i.imgur.com/AIJ7h0O.jpg)

Here is the link to the GitHub page: https://github.com/freeCodeCamp/freeCodeCamp/blob/master/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/sort-an-array-alphabetically-using-the-sort-method.english.md

**Comment by Tameshiwari** [comment]
RE:  Sort an Array Alphabetically using the sort Method Challenge...
Maybe not the cleanest but simply running this code:

`function alphabeticalOrder(arr) {`
`    // Add your code below this line`
           `return arr.sort(function(a,b) {`
`if(a < b) { return -1; }`
`if(a > b) { return 1; }`
`return 0;`
`});`
`// Add your code above this line`
`}`
`alphabeticalOrder(["a", "d", "c", "a", "z", "g"]);`

worked in my case :)

**Comment by drecali** [comment]
Hi @lasjorg 

> @drecali Not sure if I misunderstood you, by "sample code" do you mean the guide?

I was not referring to the guide, but instead to the live/production [FCC challenge page](https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/functional-programming/sort-an-array-alphabetically-using-the-sort-method) and the GitHub version. (Now that you mentioned it, I checked out the PR you linked and noticed that even in the guide, the first example for an ascending sort does not work in Chrome 70+. I'll comment on that PR as well.)

Here are screenshots and tables below summarizing the issue. In short, the sample code on the FCC site for ascending sort is an example of how to sort numbers. It works for numbers but does not work when applied to letters/strings in Chrome 70+. The sample code for the reverse alphabetical sort also doesn't work in Chrome 70+ but does in FireFox. I think it would be best to have sample code that applies to the challenge, and works across major browsers.

### FCC Live Site:
![Free Code Camp Screenshot](https://i.imgur.com/6chAVSk.jpg)

### GitHub version:
![GitHub version](https://i.imgur.com/vZInCSs.jpg)

### FCC Live Site:
| Sort Type   |      Sample Code      |  Chrome 70+ | FireFox |
|----------|:-------------:|:------:|:-------:|
| Alphabetical |  `return a - b;` | __FAIL__ | PASS
| Reverse-Alpha |  `return a < b;` |   __FAIL__ | PASS

### GitHub Version:
| Sort Type   |      Sample Code      |  Chrome 70+ | FireFox |
|----------|:-------------:|:------:|:-------:|
| Alphabetical |  `return a - b;` | __FAIL__ | PASS
| Reverse-Alpha |  `return a === b ? 0 : a < b ? 1 : -1;` __*__ |   PASS | PASS

*Note: there is a typo in the second sample code block of the GitHub version screenshot. The PR you linked addresses that as well.

Hope this info helps. Please let me know if anything is unclear.

**Comment by lasjorg** [comment]
The TLDR;

The only bug is, Chrome used to accept boolean return values like Firefox still does. After Chrome changed their sort it no longer does. This is not a bug in Chrome, but incorrect usage of the compare function in the assignment text, this has now been fixed (site is not updated yet).

The comparefn is supposed to return numeric values:

0  (a must be equal to b)
1  (a is greater than b by the ordering criterion)
-1 (a is less than b by some ordering criterion)

If the comparefn does not return numeric values it is not consistent and the sort order is implementation-defined (which is spec compliant).

@drecali

1. The ascendingOrder (return a - b) version is _not_ meant to work with strings, in any browser. It is an example of using a compare function for sorting an array of _numbers_ using subtraction.

You seem to be mixing the number (ascendingOrder) and string versions (alphabeticalOrder/reverseAlpha), not sure why.

Your screenshot and table are confusing and as far as I can tell must be wrong. The screenshot shows Chrome failing to sort an array of numbers, using ascendingOrder, which is incorrect. Then your table shows Firefox passing on an array of strings, using ascendingOrder, which it does not.

For both browsers, using ascendingOrder on an array of numbers sorts the array correctly.

Using ascendingOrder in Chrome on an array of strings leaves the array unchanged.
Using ascendingOrder in Firefox on an array of strings reverses the array but does not sort it.

Just to be clear, ascendingOrder is meant to sort the array, not reverse it. It just happens to be shown in the example text on an array of numbers that is in reverse order and so sorting the array effectively reverses it.


2. We can't put the alphabeticalOrder version in the assignment text, that is the solution to the assignment. That is why I asked.


3. For deeply nested ternaries your last style example is pretty common, however, for what we have in the assignment I think a single line expression is better. My format suggestion is based on how prettier formats the code.


Last I'm not sure if this conversation on a closed issue is considered proper etiquette or not, I simply don't have enough experience.

**Comment by ezioda004** [comment]
@drecali The above analysis is spot on. 
@lasjorg Thank you for the great explanation and its fine to have a discussion in a closed issue as long as its on topic. Otherwise you can always create new issue.

**Comment by fatemehIr** [comment]
function alphabeticalOrder(arr) {
  // Add your code below this line
  return arr.sort((a, b) => a !== b ? (a < b ? -1 : 1) : 0);
 
}
